From ac821370e18104717c582891e0a790376cf2b590 Mon Sep 17 00:00:00 2001
From: Sergey Senozhatsky <senozhatsky@chromium.org>
Date: Mon, 9 Jan 2023 12:38:36 +0900
Subject: [PATCH 2/4] zsmalloc: skip chain size calculation for pow_of_2
 classes

If a class size is power of 2 then it wastes no memory
and the best configuration is 1 physical page per-zspage.

Signed-off-by: Sergey Senozhatsky <senozhatsky@chromium.org>
---
 mm/zsmalloc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/zsmalloc.c b/mm/zsmalloc.c
index c8cd7bca4..42805e94f 100644
--- a/mm/zsmalloc.c
+++ b/mm/zsmalloc.c
@@ -2148,6 +2148,9 @@ static int calculate_zspage_chain_size(int class_size)
 	int i, min_waste = INT_MAX;
 	int chain_size = 1;
 
+	if (is_power_of_2(class_size))
+		return chain_size;
+
 	for (i = 1; i <= ZS_MAX_PAGES_PER_ZSPAGE; i++) {
 		int waste;
 
-- 
2.39.0.rc2.1.gbd5df96b79

